python/ray/serve/_private/metrics_utils.py

```diff
@@ -1,138 +1,94 @@
-        consistently metrics delivery. Python GIL will ensure that this thread gets
-        fair timeshare to execute and run.
-        """
-
-        if self.pusher_thread and self.pusher_thread.is_alive():
-            return
+    def stop_tasks(self):
+        self._stop_event.set()
 
-        def send_forever():
-            while True:
-                if self.stop_event.is_set():
-                    return
-
-                start = time.time()
-                for task in self.tasks.values():
-                    try:
-                        if start - task.last_call_succeeded_time >= task.interval_s:
-                            if task.last_ref:
-                                ready_refs, _ = ray.wait([task.last_ref], timeout=0)
-                                if len(ready_refs) == 0:
-                                    continue
-                            kwargs = task.task_func()
-                            task.last_call_succeeded_time = time.time()
-                            if task.callback_func and ray.is_initialized():
-                                task.last_ref = task.callback_func(
-                                    **kwargs, send_timestamp=time.time()
-                                )
-                    except Exception as e:
-                        logger.warning(
-                            f"MetricsPusher thread failed to run metric task: {e}"
-                        )
-
-                # For all tasks, check when the task should be executed
-                # next. Sleep until the next closest time.
-                least_interval_s = math.inf
-                if self.tasks:
-                    for task in self.tasks.values():
-                        time_until_next_push = task.interval_s - (
-                            time.time() - task.last_call_succeeded_time
-                        )
-                        least_interval_s = min(least_interval_s, time_until_next_push)
-                else:
-                    # If there are no tasks registered, fall back to the default.
-                    least_interval_s = self.NO_TASKS_REGISTERED_INTERVAL_S
-
-                time.sleep(max(least_interval_s, 0))
-
-        self.pusher_thread = threading.Thread(target=send_forever)
-        # Making this a daemon thread so it doesn't leak upon shutdown, and it
-        # doesn't need to block the replica's shutdown.
-        self.pusher_thread.setDaemon(True)
-        self.pusher_thread.start()
-
-    def __del__(self):
-        self.shutdown()
-
-    def shutdown(self):
+    async def shutdown(self):
         """Shutdown metrics pusher gracefully.
 
         This method will ensure idempotency of shutdown call.
         """
-        if not self.stop_event.is_set():
-            self.stop_event.set()
 
-        if self.pusher_thread:
-            self.pusher_thread.join()
+        self.stop_tasks()
+        if self._async_tasks:
+            await asyncio.wait(list(self._async_tasks.values()))
```

Comment:
I think instead of waiting for all tasks to complete, maybe we can just call `cancel()` on each of them? But maybe defer to @edoakes for best practice on this.

Reply:
This is the graceful shutdown path, so we should wait for them to exit